internal/utils/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,73 @@
+package ringbuffer
+
+type RingBuffer[T any] struct {
+	ring             []T
+	headPos, tailPos int
+	full             bool
```

Comment:
This might be dodgy but you could do a short ring optimisation by using `full` as signaling. Due to how the GC work this would only work with pointer free types (like stream ids as you are using it now).
You could make it a `uint8`, have `0` be false `1` be true, and values above being len in an aliased array.
Then you can use `(*T)(unsafe.Add(unsafe.Pointer(r), unsafe.Sizeof(zerovalue) * (r.full - 2)))` (note this code only work if T's size match the pointer size, slightly different expression would always work and assuming I'm not off by one)
You would setup the array as this:
```go
// Setup memory layout like this:
// type RingBuffer struct{
//   arrayPointer ----+
//   len <------------+
//  cap
//  tailPos
//  headPos
//  full (signaling for short ring optimisation)
// }
*r = RingBuffer[T]{ring: unsafe.Slice((*T)(unsafe.Add(unsafe.Pointer(r), unsafe.Sizeof(uintptr(0)))), 0, 0), full: 2}
```
As you pop element from the front you would advance the pointer in the slice:
```go
// type RingBuffer struct{
//   arrayPointer ----+
//   len              |
//  cap <-------------+
//  tailPos
//  headPos
//  full (signaling for short ring optimisation)
// }
```
You wouldn't implement a ring, for such short runs if you ever need to grow you just shift all elements back to the first aliased pointer configuration I've shown.

Also the go compiler makes assumptions about the return values from the `len` and `cap` builtins which isn't compatible with using them as aliased storage so all short operations would need to go through unsafe.

This would be very cache efficient for short ring, assuming 64 bits arch with 64 bits T you could store 4 of them this way (which is the default usage you show in framer).

Reply:
This is a usual optimisation to find in very well optimised C++ strings.

Reply:
I'm not opposed to using `unsafe` in general, but there are tradeoffs. For starters, every time we find an unexplained crash, we'll come back to the places in our code where `unsafe` is used, so it does create a maintenance burden.

As a rule of thumb, I think we should only use `unsafe` where its critical for _overall_ performance, i.e. having a direct and significant (more than 1%? 2%?) impact on transfer speed.